### PR TITLE
Feature/dx 686 add mobilepay payments

### PR DESCRIPTION
**Pull request by trieutran91**
Adding MobilePay Payments.
Finally.

### CODE REVIEW

File: src/SwedbankPay.Sdk/Payments/MobilePayPayments/MobilePayPaymentOperations.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+                        break;
+                }
+                this.Add(httpOperation.Rel, httpOperation);
+
+            }
+        }
+        public Func<PaymentAbortRequest, Task<MobilePayPaymentResponse>> Abort { get; }
+        public Func<MobilePayPaymentCancelRequest, Task<CancellationResponse>> Cancel { get; }
+        public Func<MobilePayPaymentCaptureRequest, Task<CaptureResponse>> Capture { get; }
+        public HttpOperation RedirectAuthorization { get; internal set; }
```
Comment (NullableInt) [inline src/SwedbankPay.Sdk/Payments/MobilePayPayments/MobilePayPaymentOperations.cs]:
```suggestion
        public HttpOperation RedirectAuthorization { get; }
```
Remove internal setter.

File: src/SwedbankPay.Sdk/Payments/MobilePayPayments/MobilePayPaymentAuthorization.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+{
+    public class MobilePayPaymentAuthorization : IdLink
+    {
+        public MobilePayPaymentAuthorization(
+                             string maskedPan,
+                             string expiryDate,
+                             string panToken,
+                             string cardBrand,
+                             string cardType,
+                             string issuingBank,
```
Comment (NullableInt) [inline src/SwedbankPay.Sdk/Payments/MobilePayPayments/MobilePayPaymentAuthorization.cs]: Parameter `issuingBank` is not in use

File: src/Samples/Sample.AspNetCore.SystemTests/Test/PaymentTests/Base/PaymentTests.cs [hunk 3]
```diff
@@ -60,7 +61,7 @@ public void OneTimeSetUp()
                 httpClient,
                 new PaymentOrdersResource(httpClient),
                 new ConsumersResource(httpClient),
-                new PaymentsResource(httpClient, new CardPaymentsResource(httpClient), new SwishPaymentsResource(httpClient)));
+                new PaymentsResource(httpClient, new CardPaymentsResource(httpClient), new SwishPaymentsResource(httpClient), new MobilePayPaymentsResource(httpClient)));
```
Comment (asbjornu) [inline src/Samples/Sample.AspNetCore.SystemTests/Test/PaymentTests/Base/PaymentTests.cs]: If the default implementations are used, `PaymentsResource` should be able to construct them itself.

```suggestion
                new PaymentsResource(httpClient));
```

File: src/SwedbankPay.Sdk.Tests/TestBuilders/PaymentRequestBuilder.cs [hunk 4]
```diff
@@ -56,6 +56,19 @@ public SwishPaymentRequest BuildSwishPaymentRequest()
             );
         }
 
+        public Payments.MobilePayPayments.MobilePayPaymentRequest BuildMobilePayRequest()
+        {
+            return new Payments.MobilePayPayments.MobilePayPaymentRequest(
```
Comment (asbjornu) [inline src/SwedbankPay.Sdk.Tests/TestBuilders/PaymentRequestBuilder.cs]:
```suggestion
            return new MobilePayPaymentRequest(
```

File: src/SwedbankPay.Sdk/Payments/MobilePayPayments/MobilePayPaymentOperations.cs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        public Func<MobilePayPaymentReversalRequest, Task<ReversalResponse>> Reversal { get; }
```
Comment (asbjornu) [inline src/SwedbankPay.Sdk/Payments/MobilePayPayments/MobilePayPaymentOperations.cs]:
```suggestion
        public Func<MobilePayPaymentReversalRequest, Task<ReversalResponse>> Reverse { get; }
```

File: src/SwedbankPay.Sdk/Payments/MobilePayPayments/MobilePayPaymentResponse.cs [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+            UserAgent = userAgent;
+            Language = language;
+            Urls = urls;
+            PayeeInfo = payeeInfo;
+            Transactions = transactions;
+            Authorizations = authorizations;
+            Captures = captures;
+            Reversals = reversals;
+            Cancellations = cancellations;
+            MetaData = metaData;
```
Comment (asbjornu) [inline src/SwedbankPay.Sdk/Payments/MobilePayPayments/MobilePayPaymentResponse.cs]: `metadata` is written with a lowercase `d` in the API, so please make that consistent throughout the SDK as well. Perhaps not in this PR, but in a separate one.

```suggestion
            Metadata = metadata;
```

Comment (trieutran91) [reply src/SwedbankPay.Sdk/Payments/MobilePayPayments/MobilePayPaymentResponse.cs]: It seems like uppercase `d` is used in the SDK. I'll make a seperate Jira-task for this.

File: src/Samples/Sample.AspNetCore.SystemTests/Test/PaymentTests/Base/PaymentTests.cs [hunk 7]
```diff
@@ -56,7 +57,11 @@ public void OneTimeSetUp()
             httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", Environment.GetEnvironmentVariable("Payex.Api.Token", EnvironmentVariableTarget.User));
             #endif
 
-            SwedbankPayClient = new SwedbankPayClient(httpClient);
+            SwedbankPayClient = new SwedbankPayClient(
+                httpClient,
+                new PaymentOrdersResource(httpClient),
+                new ConsumersResource(httpClient),
+                new PaymentsResource(httpClient));
```
Comment (NullableInt) [inline src/Samples/Sample.AspNetCore.SystemTests/Test/PaymentTests/Base/PaymentTests.cs]: You can undo this change.

File: src/Samples/Sample.AspNetCore.SystemTests/Test/PaymentTests/Base/PaymentTests.cs [hunk 8]
```diff
@@ -24,6 +24,7 @@
 using SwedbankPay.Sdk.Payments;
 using SwedbankPay.Sdk.Payments.CardPayments;
 using SwedbankPay.Sdk.Payments.SwishPayments;
+using SwedbankPay.Sdk.Payments.MobilePayPayments;
```
Comment (NullableInt) [inline src/Samples/Sample.AspNetCore.SystemTests/Test/PaymentTests/Base/PaymentTests.cs]: Remove unused usings, There are 6 here.

File: src/SwedbankPay.Sdk.Tests/PaymentTests/MobilePayTests.cs [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+        [Fact]
+        public async Task CreateMobilePayPayment_ShouldReturnPayment()
+        {
+            var paymentRequest = this.paymentRequestBuilder.WithMobilePayTestValues(this.payeeId).BuildMobilePayRequest();
+            var mobilePayPayment = await this.Sut.Payments.MobilePayPayments.Create(paymentRequest, PaymentExpand.All);
+            Assert.NotNull(mobilePayPayment);
+        }
+
+        [Fact]
+        public async Task GetPayment()
```
Comment (NullableInt) [inline src/SwedbankPay.Sdk.Tests/PaymentTests/MobilePayTests.cs]: What does this test?

File: src/SwedbankPay.Sdk.Tests/ResourceTestsBase.cs [hunk 10]
```diff
@@ -7,6 +7,7 @@
 using SwedbankPay.Sdk.Payments;
 using SwedbankPay.Sdk.Payments.CardPayments;
 using SwedbankPay.Sdk.Payments.SwishPayments;
+using SwedbankPay.Sdk.Payments.MobilePayPayments;
```
Comment (NullableInt) [inline src/SwedbankPay.Sdk.Tests/ResourceTestsBase.cs]: This using is not required.

File: src/SwedbankPay.Sdk.Tests/TestBuilders/PaymentRequestBuilder.cs [hunk 11]
```diff
@@ -56,6 +58,20 @@ public SwishPaymentRequest BuildSwishPaymentRequest()
             );
         }
 
+        public Payments.MobilePayPayments.MobilePayPaymentRequest BuildMobilePayRequest()
```
Comment (NullableInt) [inline src/SwedbankPay.Sdk.Tests/TestBuilders/PaymentRequestBuilder.cs]:
```suggestion
        public MobilePayPaymentRequest BuildMobilePayRequest()
```

File: src/SwedbankPay.Sdk/AssemblyInfo.cs [hunk 12]
```diff
@@ -0,0 +1,3 @@
+﻿using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("SwedbankPay.Sdk.Tests")]
```
Comment (NullableInt) [inline src/SwedbankPay.Sdk/AssemblyInfo.cs]: File is not needed

File: src/SwedbankPay.Sdk/Payments/MobilePayPayments/MobilePayPaymentAuthorizationTransaction.cs [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+            FailedActivityName = failedActivityName;
+            FailedErrorCode = failedErrorCode;
+            FailedErrorDescription = failedErrorDescription;
+            IsOperational = isOperational;
+            Operations = operations;
+        }
+
+
+        public Amount Amount { get; }
+
```
Comment (NullableInt) [inline src/SwedbankPay.Sdk/Payments/MobilePayPayments/MobilePayPaymentAuthorizationTransaction.cs]: Remove unnecessary whitespace.

File: src/SwedbankPay.Sdk/Payments/MobilePayPayments/MobilePayPaymentAuthorizationsResponse.cs [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+    {
+        public MobilePayPaymentAuthorizationsResponse(Uri payment, MobilePayPaymentAuthorizationListResponse authorizationList)
+        {
+            Payment = payment;
+            AuthorizationList = authorizationList;
+        }
+
+
+        public MobilePayPaymentAuthorizationListResponse AuthorizationList { get; }
+
```
Comment (NullableInt) [inline src/SwedbankPay.Sdk/Payments/MobilePayPayments/MobilePayPaymentAuthorizationsResponse.cs]: Remove unnecessary whitespace.
```suggestion
```

File: src/SwedbankPay.Sdk/Payments/MobilePayPayments/MobilePayPaymentCancelRequest.cs [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+        {
+            protected internal CancelTransaction(string payeeReference, string description)
+            {
+                PayeeReference = payeeReference;
+                Description = description;
+            }
+
+
+            public string Description { get; }
+
```
Comment (NullableInt) [inline src/SwedbankPay.Sdk/Payments/MobilePayPayments/MobilePayPaymentCancelRequest.cs]:
```suggestion
```

File: src/SwedbankPay.Sdk/Payments/MobilePayPayments/MobilePayPaymentRequest.cs [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+            public Dictionary<string, object> MetaData { get; }
+            public Operation Operation { get; set; }
+            public PayeeInfo PayeeInfo { get; internal set; }
+            public string PayerReference { get; set; }
+            public List<Price> Prices { get; set; }
+            public Urls Urls { get; }
+            public string UserAgent { get; set; }
+            public PrefillInfo PrefillInfo { get; set; }
+        }
+        public class MobilePayPaymentRequestObject
```
Comment (NullableInt) [inline src/SwedbankPay.Sdk/Payments/MobilePayPayments/MobilePayPaymentRequest.cs]:
```suggestion


        public class MobilePayPaymentRequestObject
```

File: src/SwedbankPay.Sdk/Payments/MobilePayPayments/MobilePayPaymentRequest.cs [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+            {
+                ShoplogoUrl = shoplogoUrl;
+            }
+
+            /// <summary>
+            ///    URI to logo that will be visible at MobilePay payment menu
+            /// </summary>
+            public Uri ShoplogoUrl { get; set; }
+
+
```
Comment (NullableInt) [inline src/SwedbankPay.Sdk/Payments/MobilePayPayments/MobilePayPaymentRequest.cs]:
```suggestion
```

File: src/SwedbankPay.Sdk/Payments/MobilePayPayments/MobilePayPaymentRequest.cs [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+            protected internal MobilePayPaymentRequestObject(Uri shoplogoUrl)
+            {
+                ShoplogoUrl = shoplogoUrl;
+            }
+
+            /// <summary>
+            ///    URI to logo that will be visible at MobilePay payment menu
+            /// </summary>
+            public Uri ShoplogoUrl { get; set; }
+
```
Comment (NullableInt) [inline src/SwedbankPay.Sdk/Payments/MobilePayPayments/MobilePayPaymentRequest.cs]:
```suggestion
```

File: src/SwedbankPay.Sdk.Tests/PaymentTests/MobilePayTests.cs [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+        [Fact]
+        public async Task CreateMobilePayPayment_ShouldReturnPayment()
+        {
+            var paymentRequest = this.paymentRequestBuilder.WithMobilePayTestValues(this.payeeId).BuildMobilePayRequest();
+            var mobilePayPayment = await this.Sut.Payments.MobilePayPayments.Create(paymentRequest, PaymentExpand.All);
+            Assert.NotNull(mobilePayPayment);
+        }
+
+        [Fact]
+        public async Task GetPayment_CheckIfGetFunctionWorks()
```
Comment (NullableInt) [inline src/SwedbankPay.Sdk.Tests/PaymentTests/MobilePayTests.cs]: This test does nothing, might as well remove?

Comment (trieutran91) [reply src/SwedbankPay.Sdk.Tests/PaymentTests/MobilePayTests.cs]: It's gone!

**Comment by asbjornu** [comment]
@zunkas, can you please investigate why `sdk-dotnet-Dev-CI` fails?

File: src/SwedbankPay.Sdk/Payments/MobilePayPayments/MobilePayPayment.cs [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+            var operations = new MobilePayPaymentOperations(paymentResponse.Operations, client);
+            Operations = operations;
+        }
+
+        public MobilePayPaymentOperations Operations { get; }
+
+        public PaymentResponseObject PaymentResponse { get; }
+
+
+        internal static async Task<MobilePayPayment> Create(MobilePayPaymentRequest paymentRequest,
```
Comment (asbjornu) [inline src/SwedbankPay.Sdk/Payments/MobilePayPayments/MobilePayPayment.cs]: The `Create` method should be exposed on the `MobilePayPaymentsResource` directly under `SwedbankPayClient`, no?

File: src/SwedbankPay.Sdk/Payments/MobilePayPayments/MobilePayPayment.cs [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+                                                   string paymentExpand)
+        {
+            var url = new Uri($"/psp/mobilepay/payments{paymentExpand}", UriKind.Relative);
+
+            var paymentResponse = await client.PostAsJsonAsync<MobilePayPaymentResponse>(url, paymentRequest);
+            return new MobilePayPayment(paymentResponse, client);
+        }
+
+
+        internal static async Task<MobilePayPayment> Get(Uri id, HttpClient client, string paymentExpand)
```
Comment (asbjornu) [inline src/SwedbankPay.Sdk/Payments/MobilePayPayments/MobilePayPayment.cs]: The same applies to `Get` as to `Create`.

File: src/SwedbankPay.Sdk/Payments/MobilePayPayments/MobilePayPayment.cs (hunk 20 above)
Comment (trieutran91) [reply src/SwedbankPay.Sdk/Payments/MobilePayPayments/MobilePayPayment.cs]: Isn't this method exposed through `MobilePayPaymentsResource` under `SwedbankPayClient`?

Comment (asbjornu) [reply src/SwedbankPay.Sdk/Payments/MobilePayPayments/MobilePayPayment.cs]: It may be, but I think the implementation is in the wrong place. Is this how `Create` is implemented in all other payment instruments as well?

**Comment by trieutran91** [comment]
Yes, both `Create` and `Get` is implemented in the same way in the other instruments. 
Do you recommend to change this implementation in these PR's or should we handle it in a different task/PR?
